### PR TITLE
IGNITE-17679 Consistency check/repair and Read Repair documentation update

### PR DESCRIPTION
**Pull request by anton-vinogradov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: docs/_docs/key-value-api/read-repair.adoc [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+* Near caches.
+* Caches that use "read-through" mode.
+====
+
+
+`Read Repair` refers to a technique of repairing inconsistencies between primary and backup copies of data during normal read operations. When a specific key (or keys) is read by a user operation, Ignite checks the values for the given key in all backup copies.
+
+The `Read Repair` mode is designed to maintain consistency. However, read operations become {tilde}2 times more costly because backup copies are checked in addition to the primary copy. It is generally not advisable to use this mode all the time, but rather on a once-in-a-while basis.
+
+To enable the `Read Repair` mode, obtain an instance of the cache that enables Read Repair reads as follows:
```
Comment (ivandasch) [inline docs/_docs/key-value-api/read-repair.adoc:35]: Sounds too prescriptive

`it is advisable to obtain` sounds much friendlier.

File: docs/_docs/key-value-api/read-repair.adoc [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+May cause IgniteException when the fix is impossible (unable to detect the newest entry)::
+* Null(s) found as well as non-null values for the same key. Null (missed entry) has no version, so, it can not be compared with the versioned entry.
+* Entries with the same version have different values.
+| `PRIMARY` | Value from the primary node wins.
+| `RELATIVE_MAJORITY` a| The relative majority, any value found more times than any other wins.
+Works for an even number of copies (which is typical of Ignite) instead of an absolute majority.
+
+May cause IgniteException when unable to detect values found more times than any other.
+
```
Comment (ivandasch) [inline docs/_docs/key-value-api/read-repair.adoc:58]: `when it is unable to detect`

File: docs/_docs/key-value-api/read-repair.adoc [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+For example, when we have 5 copies (4 backups)::
+* and value `A` found twice, but `X`, `Y` and `Z` only once, `A` wins,
```
Comment (ivandasch) [inline docs/_docs/key-value-api/read-repair.adoc:60]: `when 5 copies (4 backups) are given`

File: docs/_docs/key-value-api/read-repair.adoc [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+* but, when `A` is found twice, as well as `B`, and `X` only once, the strategy is unable to determine the winner.
+
+When we have 4 copies (3 backups), any value found two or more times, when others are found only once, is the winner.
+| `REMOVE` | Inconsistent entries will be removed.
+| `CHECK_ONLY` | Only check will be performed.
+|===
+
+== Events
+link:https://ignite.apache.org/releases/{version}/javadoc/org/apache/ignite/events/EventType.html#EVT_CONSISTENCY_VIOLATION[Сonsistency Violation Event] will be recorded for each violation in case it's configured as recordable. You can listen to this event to get notified about inconsistency issues.
```
Comment (ivandasch) [inline docs/_docs/key-value-api/read-repair.adoc:69]: `You may listen to this event`

File: docs/_docs/key-value-api/read-repair.adoc [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+Values will be repaired::
+* automatically for transactions that have `TransactionConcurrency.OPTIMISTIC` concurrency mode or `TransactionIsolation.READ_COMMITTED` isolation level,
+
+* at commit() phase for transactions that have `TransactionConcurrency.PESSIMISTIC` concurrency mode and isolation level other than `TransactionIsolation.READ_COMMITTED`
+
+[WARNING]
+====
+[discrete]
+=== Limitations
+This proxy usage does not guarantee "all copies check" in case the value is already cached inside the transaction.
```
Comment (ivandasch) [inline docs/_docs/key-value-api/read-repair.adoc:83]: `have been already cached`

File: docs/_docs/key-value-api/read-repair.adoc [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+In case you do use not a `READ_COMMITTED` isolation mode and already have a cached value, for example already read the value or performed a write, you'll just gain the cached value.
```
Comment (ivandasch) [inline docs/_docs/key-value-api/read-repair.adoc:85]: Too many `already`
`If you don't use a READ_COMMITTED`
`have already read the value or performed a write`
`you will just get` (not sure what you meant here)

File: docs/_docs/tools/control-script.adoc [hunk 7]
```diff
@@ -507,6 +511,102 @@ Partition instances: [PartitionHashRecord [isPrimary=true, partHash=97595430, up
...
+===  Additional configuration required
+The command uses special link:https://ignite.apache.org/releases/{version}/javadoc/org/apache/ignite/events/EventType.html#EVT_CONSISTENCY_VIOLATION[Consistency Violation Event] to detect the consistency violations. This event must be enabled before the command execution.
+
+Refer to the link:events/listening-to-events#enabling-events[Enabling Events] section for details.
+====
+
+`idle_verify` command provides the inconsistent cache group names and partitions list as a result.
+The `repair` command allows performing cache consistency check and repair (when possible) using the link:key-value-api/read-repair[Read Repair] approach for every inconsistent partition found by `idle_verify`.
+
+The command uses special strategies to perform the repair. It's recommended to use `CHECK_ONLY` strategy to list inconsistent values and then chose the proper link:key-value-api/read-repair#strategies[Repair Strategy].
```
Comment (ivandasch) [inline docs/_docs/tools/control-script.adoc:536]: `choose`

File: docs/_docs/tools/control-script.adoc [hunk 8]
```diff
@@ -507,6 +511,102 @@ Partition instances: [PartitionHashRecord [isPrimary=true, partHash=97595430, up
...
+
+By default, found inconsistent entries will be listed in the application log, you may change the location by configuring the logger for a special logging path for the `org.apache.ignite.internal.visor.consistency` package.
```
Comment (ivandasch) [inline docs/_docs/tools/control-script.adoc:538]: `log. You may`

File: docs/_docs/key-value-api/read-repair.adoc (hunk 1 above)
Comment (anton-vinogradov) [reply docs/_docs/key-value-api/read-repair.adoc:35]: It's not advisable, it is necessary

File: docs/_docs/key-value-api/read-repair.adoc (hunk 6 above)
Comment (anton-vinogradov) [reply docs/_docs/key-value-api/read-repair.adoc:85]: it means you'll gain a cached value

File: docs/_docs/tools/control-script.adoc [hunk 9]
```diff
@@ -460,24 +460,28 @@ control.sh --cache reset_lost_partitions cacheName1,cacheName2,...
...
+
+Finally, consistency can be repaired if necessary.
 
 Let's review these usage scenarios in more detail.
 
 === Verifying Partition Checksums
 
-//Even if update counters and size are equal on the primary and backup nodes, there might be a case when the primary and backup  diverge due to some critical failure.
+Even if update counters and size are equal on the primary and backup nodes, there might be a case when the primary and backup diverge due to some critical failure.
+
```
[the remark this answers is not in the document]
Comment (anton-vinogradov) [reply docs/_docs/tools/control-script.adoc:480]: `might` already used some words before

**Review by ivandasch — APPROVED** [review]

**Pull request merged**